compiler/rustc_expand/src/mbe/diagnostics.rs

```diff
@@ -181,6 +184,19 @@ impl<'a, 'cx> CollectTrackerAndEmitter<'a, 'cx, '_> {
     }
 }
 
+fn maybe_annotate_mismatch(
+    error: &mut DiagnosticBuilder<'_, ErrorGuaranteed>,
+    found: &Token,
+    expected: &MatcherLoc,
+) {
+    if let MatcherLoc::Token { token: expected_token } = expected
+        // The tokens are printed the same - there's a nonterminal at play. This isn't perfect but good enough.
+        && pprust::token_to_string(expected_token) == pprust::token_to_string(found)
```

Comment:
Why not check that the expected token is a non-terminal directly?
As for pretty-printing, I'm not even sure why it works.

Reply:
That sounds good. I'm checking both for nonterminals because I've also had nonterminals in the matcher before (generating a macro_rules from a macro_rules)